pkg/cmd/cloudenv/options.go

```diff
@@ -0,0 +1,374 @@
+	"path/filepath"
+	"runtime"
+	"text/template"
+
+	sprigv3 "github.com/Masterminds/sprig/v3"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/gardener/gardenctl-v2/internal/gardenclient"
+	"github.com/gardener/gardenctl-v2/internal/util"
+	"github.com/gardener/gardenctl-v2/pkg/cmd/base"
+	"github.com/gardener/gardenctl-v2/pkg/target"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+)
+
+//go:embed templates
+var fsys embed.FS
+
+var (
+	ErrNoShootTargeted               = errors.New("no shoot cluster targeted")
+	ErrNeitherProjectNorSeedTargeted = errors.New("neither project nor seed is targeted")
+	ErrProjectAndSeedTargeted        = errors.New("project and seed must not be targeted at the same time")
+)
+
+type cmdOptions struct {
+	base.Options
+
+	// Unset resets environment variables and configuration of the cloudprovider CLI for your shell.
+	Unset bool
+	// Shell to configure.
+	Shell string
+	// GardenDir is the configuration directory of gardenctl.
+	GardenDir string
+	// CmdPath is the path of the called command.
+	CmdPath string
+	// CurrentTarget is pointing to the current target
+	CurrentTarget target.Target
+}
+
+var _ base.CommandOptions = &cmdOptions{}
+
+// Complete adapts from the command line args to the data required.
+func (o *cmdOptions) Complete(f util.Factory, cmd *cobra.Command, _ []string) error {
+	o.GardenDir = f.GardenHomeDir()
+	o.Shell = cmd.Name()
+	o.CmdPath = cmd.Parent().CommandPath()
+
+	return nil
+}
+
+// Validate validates the provided command options.
+func (o *cmdOptions) Validate() error {
+	if o.Shell == "" {
+		return pflag.ErrHelp
+	}
+
+	s := Shell(o.Shell)
+	if err := s.Validate(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Run does the actual work of the command.
+func (o *cmdOptions) Run(f util.Factory) error {
+	// target manager
+	m, err := f.Manager()
+	if err != nil {
+		return err
+	}
+
+	// current target
+	o.CurrentTarget, err = m.CurrentTarget()
+	if err != nil {
+		return err
+	}
+
+	gardenName := o.CurrentTarget.GardenName()
+	projectName := o.CurrentTarget.ProjectName()
+	seedName := o.CurrentTarget.SeedName()
+	shootName := o.CurrentTarget.ShootName()
+
+	// validate current target
+	if shootName == "" {
+		return ErrNoShootTargeted
+	} else if projectName == "" && seedName == "" {
+		return ErrNeitherProjectNorSeedTargeted
+	} else if projectName != "" && seedName != "" {
+		return ErrProjectAndSeedTargeted
+	}
+
+	// garden client
+	gardenClient, err := m.GardenClient(gardenName)
+	if err != nil {
+		return fmt.Errorf("failed to create garden cluster client: %w", err)
+	}
+
+	ctx := f.Context()
+
+	var shoot *gardencorev1beta1.Shoot
+
+	if projectName != "" {
+		shoot, err = gardenClient.GetShootByProject(ctx, projectName, shootName)
+		if err != nil {
+			return err
+		}
+	} else {
+		shoot, err = gardenClient.GetShootBySeed(ctx, seedName, shootName)
+		if err != nil {
+			return err
+		}
+	}
```

Comment:
I think the implementation of the FindShoot method should not be in the GardenClient because it is target specific. Wether we have some kind of query language which allows combining match patterns with `or` or we should move this method back to the manager. But I think the two cases targeted by seed or targeted by project should not be handled in the client. I would recommend to store the namespace for a project in the target. This would also be good for performance reasons.

Reply:
✔️